### PR TITLE
#8751: automatically refresh spreadsheet list when activating a mod and configuring a new google auth config

### LINKED ISSUE #8751: Google Sheet selector widget not refreshing after creating a Google configuration
## Describe the bug

Google Sheet selector widget not refreshing after creating/selecting Google configuration

## To Reproduce

1. Click to activate a mod using Google sheets
2. Create a new Google configuration

## Actual behavior

Google Sheet selector doesn't populate after selecting the Google configuration

## Expected behavior

Google Sheet selector refreshes when you select a Google configuration and/or change the Google Configuration

## Workaround

- Refresh the page

## Related Code

- https://github.com/pixiebrix/pixiebrix-extension/blob/72334200b17050bde676dbf234650f147f53ac8a/src/contrib/google/sheets/ui/SpreadsheetPickerWidget.tsx#L40-L40
- https://github.com/pixiebrix/pixiebrix-extension/blob/72334200b17050bde676dbf234650f147f53ac8a/src/contrib/google/sheets/core/useGoogleAccount.ts#L32-L32

## Discussion

- Is useGoogleAccount still returning null after the configuration is created?

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Fixes #8751 

## Demo
https://www.loom.com/share/415b893d9ef44ce1b314e1e6675c7581

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

**Review by BLoe — APPROVED** [review]

File: src/contrib/google/sheets/core/useGoogleAccount.ts [hunk 1]
```diff
@@ -19,24 +19,33 @@ import { type SanitizedIntegrationConfig } from "@/integrations/integrationTypes
...
 import ModIntegrationsContext from "@/mods/ModIntegrationsContext";
 import { validateRegistryId } from "@/types/helpers";
 import reportError from "@/telemetry/reportError";
+import { ReusableAbortController } from "abort-utils";
+import { oauth2Storage } from "@/auth/authConstants";
+import { isEmpty } from "lodash";
+import useOnMountOnly from "@/hooks/useOnMountOnly";
 
 const GOOGLE_PKCE_INTEGRATION_ID = validateRegistryId("google/oauth2-pkce");
+const loginController = new ReusableAbortController();
```
Comment (twschiller) [inline src/contrib/google/sheets/core/useGoogleAccount.ts:32]: Any reason this is a module-level variable vs. a useState in the component?

File: src/contrib/google/sheets/core/useGoogleAccount.ts [hunk 2]
```diff
@@ -51,6 +60,20 @@ function useGoogleAccount(): FetchableAsyncState<SanitizedIntegrationConfig | nu
...
+  useEffect(() => {
+    // Automatically refetch the google account on login
+    oauth2Storage.onChanged((newValue) => {
+      const { id } = googleAccountAsyncState.data ?? {};
+      // eslint-disable-next-line security/detect-object-injection -- not user provided
+      if (id && !isEmpty(newValue[id])) {
+        googleAccountAsyncState.refetch();
+        loginController.abortAndReset();
+      }
+    }, loginController.signal);
```
Comment (twschiller) [inline src/contrib/google/sheets/core/useGoogleAccount.ts:73]: Won't this add onChanged event handler for each transition in useAsyncState (isFetching, isSuccess, etc.)

Comment (twschiller) [reply src/contrib/google/sheets/core/useGoogleAccount.ts:73]: Should the useEffect return value remove the onChanged listener?

File: src/contrib/google/sheets/core/useGoogleAccount.ts [hunk 3]
```diff
@@ -51,6 +60,20 @@ function useGoogleAccount(): FetchableAsyncState<SanitizedIntegrationConfig | nu
       return null;
     }
   }, [googleDependency]);
+
+  useEffect(() => {
+    // Automatically refetch the google account on login
+    oauth2Storage.onChanged((newValue) => {
+      const { id } = googleAccountAsyncState.data ?? {};
+      // eslint-disable-next-line security/detect-object-injection -- not user provided
+      if (id && !isEmpty(newValue[id])) {
```
Comment (twschiller) [inline src/contrib/google/sheets/core/useGoogleAccount.ts:69]: Do we need it to refetch if the user "x"'s out of the configuration so the sheet selector will reset?

File: src/contrib/google/sheets/core/useGoogleAccount.ts (hunk 2 above)
Comment (twschiller) [reply src/contrib/google/sheets/core/useGoogleAccount.ts:73]: I'd recommend extracting this useEffect hook out to a helper hook given that it involves that useOnMountOnly and a module-level variable too

File: src/contrib/google/sheets/core/useGoogleAccount.ts (hunk 1 above)
Comment (grahamlangford) [reply src/contrib/google/sheets/core/useGoogleAccount.ts:32]: Just managing referential equality. I could add it to a useState, but managing it as a module-level variable is consistent with other usages

File: src/contrib/google/sheets/core/useGoogleAccount.ts (hunk 2 above)
Comment (grahamlangford) [reply src/contrib/google/sheets/core/useGoogleAccount.ts:73]: > Won't this add onChanged event handler for each transition in useAsyncState (isFetching, isSuccess, etc.)

Technically yes. I can limit it to only when the data changes

Comment (twschiller) [reply src/contrib/google/sheets/core/useGoogleAccount.ts]: > Technically yes. I can limit it to only when the data changes

I don't think you need to do that -- you'd just need to unregister the event handler in the useEffect unmount return value

File: src/contrib/google/sheets/core/useGoogleAccount.ts [hunk 4]
```diff
@@ -26,18 +26,35 @@ import reportError from "@/telemetry/reportError";
...
+      const { id } = googleAccount ?? {};
+      // eslint-disable-next-line security/detect-object-injection -- not user provided
+      if (id && !isEmpty(newValue[id])) {
+        refetch();
+        loginController.abortAndReset();
+      }
+    }, loginController.signal);
+
+    return () => {
+      loginController.abortAndReset();
```
Comment (twschiller) [inline src/contrib/google/sheets/core/useGoogleAccount.ts:49]: I think this break things because it aborts at a module-level.

Comment (grahamlangford) [reply src/contrib/google/sheets/core/useGoogleAccount.ts:49]: Per our discussion in slack, moving the loginController into the useEffect in a separate PR

**Pull request merged**